truelearn/preprocessing/wikifier.py

```diff
@@ -0,0 +1,130 @@
+from typing import Optional, Union
+from urllib import request
+from urllib import parse
+from urllib import error
+
+import orjson
+
+Annotation = dict[str, Union[str, float, None]]
+WikifierResponse = dict[str, Union[list[str], list[dict]]]
+
+
+class Wikifier:
+    """Used to make requests to the Wikifier API.
+
+    Attributes:
+        api_key: string representing the API key needed to make the
+          request, get one from https://wikifier.org/register.html
+
+    .. [1] Janez Brank, Gregor Leban, Marko Grobelnik. Annotating Documents
+       with Relevant Wikipedia Concepts. Proceedings of the Slovenian Conference
+       on Data Mining and Data Warehouses (SiKDD 2017), Ljubljana, Slovenia,
+       9 October 2017.
+    """
+
+    def __init__(self, api_key: str) -> None:
+        """Inits Wikifier class with api_key."""
+        if isinstance(api_key, str):
+            self.api_key = api_key
+        else:
+            raise TypeError("api_key should be a string")   
+
+    def wikify(
+            self, text: str, df_ignore: int = 50,
+            words_ignore: int = 50, top_n: Optional[int] = None
+        ) -> list[Annotation]:
+        """Annotates input text using the Wikifier API.
+
+        Args:
+            text: string representing the text to annotate.
+            df_ignore: int representing the nTopDfValuesToIgnore value from
+              the Wikifier API, used to ignore frequently-occurring words.
+            words_ignore: int representing the nWordsToIgnoreFromList from the
+              Wikifier API, also used to ignore frequently-occurring words.
+            top_n: the number of annotations to return, e.g. top_n = 5 would
+              only return the top 5 annotations by pageRank.
+
+        Returns:
+            The list of annotations obtained from the Wikifier API,
+            or nothing if an exception is raised.
+
+        Raises:
+            ValueError: The response from Wikifier contained an error message
+              or the API key is not valid.
+            RuntimeError: The HTTP request returns a status code representing
+              an error
+        """
+        try:
+            resp = self.__make_wikifier_request(
+                text, df_ignore, words_ignore
+            )
+        except error.HTTPError as err:
+            raise RuntimeError(str(err)) from err
+

```

Comment:
@yuxqiu @KD-7 do you have any thoughts on this? I'm raising a RuntimeError from urllib's HTTPError so that the user does not have to import urllib to handle it, which seems like it would be a bad practice. Is there a better way to do this?

Reply:
You marked it as resolved, you could handle the 404 error if you wanted to

Reply:
I'm handling the error by chaining it into a different one that the user can deal with, I don't know what else I would do to handle it, besides perhaps adding more information to the Runtime error before raising it

Reply:
> I'm handling the error by chaining it into a different one that the user can deal with, I don't know what else I would do to handle it, besides perhaps adding more information to the Runtime error before raising it

see my updated comment

Reply:
You mean except the HTTPError in the __make_wikifier_request method rather than in wikify? Will do